### PR TITLE
Refactor result handling in training loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Minor refactor, restores a tiny bit of sanity.

Removes the following state on TrainLoop:
- `_cur_step_result`
- `_cur_grad_norm_dic`

Fixes typo in spelling of `grad_norm_dict`

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -703,11 +702,12 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
                     with self.block_ddp_sync_behaviour():
-                        self.training_step_and_backward(
+                        result = self.training_step_and_backward(
                             split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
                         )
 
                     batch_outputs = self._process_closure_result(
+                        opt_closure_result=result,
                         batch_outputs=batch_outputs,
                         opt_idx=opt_idx,
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:712]: If i'm not mistaken this batch_outputs and the one in L741-L745 can now be extracted one indentation level higher

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: good observation. almost, but had to change  `self.update_running_loss(` slightly, because it assumed process_result to run first to update the accumulated loss. 

It makes it slightly harder to review here, but should help with future refactors

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -195,14 +195,14 @@ def cache_training_step_metrics(self, opt_closure_result):
...
-    def log_metrics(self, metrics, grad_norm_dic, step=None):
+    def log_metrics(self, metrics, grad_norm_dict, step=None):
         """Logs the metric dict passed in.
         If `step` parameter is None and `step` key is presented is metrics,
         uses metrics["step"] as a step
 
         Args:
             metrics (dict): Metric values
-            grad_norm_dic (dict): Gradient norms
+            grad_norm_dict (dict): Gradient norms
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:205]: yes, logger connector will be refactored
the only reason my PR touches this class is because I auto-renamed the variable everywhere.

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -668,19 +666,19 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
             self.warning_cache.warn("train_dataloader yielded None. If this was on purpose, ignore this warning...")
             return AttributeDict(
                 signal=0,
-                grad_norm_dic=grad_norm_dic,
+                grad_norm_dict={},
                 training_step_output_for_epoch_end=batch_outputs,
             )
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:671]: yes good idea and attribute dict has to change as part of our refactor efforts anyway

**Review by ananthsub — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
